---
bibliography: ../ref.bib
---

# Available references

```{r, echo=FALSE}
knitr::opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE, cache=TRUE)
library(BiocStyle)
```

## Overview

`r Biocpkg("SingleR")` provides access to several reference datasets through convenient retrieval functions.
Most of these references are derived from bulk RNA-seq or microarray data of cell populations
that (hopefully) consist of a pure cell type after sorting and/or culturing.
Each dataset contains a log-normalized expression matrix that is intended to be comparable 
to log-UMI counts from common single-cell protocols,
typcially after adjusting bulk RNA-seq data for differences in gene length.
It also provides three levels of cell type annotation in its column metadata:

- `label.main`, broad annotation that defines the major cell types.
This has few unique levels that allows for fast annotation but at low resolution.
- `label.fine`, fine-grained annotation that defines subtypes or states.
This has more unique levels that results in slower annotation but at much higher resolution.
- `label.ont`, fine-grained annotation mapped to the standard vocabulary in the [Cell Ontology](https://www.ebi.ac.uk/ols/ontologies/cl).
This enables synchronization of labels across references as well as dynamic adjustment of the resolution,
see Chapter \@ref(Harmonizing-labels) for more details.

By default, gene annotation is returned in terms of gene symbols,
but they can be coerced to Ensembl annotation with `ensembl=TRUE` for more robust cross-referencing across studies.
More details for each dataset can be viewed on the corresponding help page for its retrieval function (e.g., `?ImmGenData`).

## General-purpose references

### Human primary cell atlas (HPCA)

@hpcaRef
  
```{r, echo=FALSE, message=FALSE}
library(SingleR)
ref <- HumanPrimaryCellAtlasData()
samples <- colData(ref)[,c("label.main", "label.fine","label.ont")]
samples <- as.data.frame(samples)
DT::datatable(unique(samples))
```

### Blueprint/ENCODE

Blueprint [@blueprintRef] and Encode [@encodeRef],

```{r, echo=FALSE, message=FALSE}
ref <- BlueprintEncodeData()
samples <- colData(ref)[,c("label.main", "label.fine","label.ont")]
samples <- as.data.frame(samples)
DT::datatable(unique(samples))
```

### Mouse RNA-seq

A collection of mouse data sets downloaded from GEO [@Benayoun2019].

```{r, echo=FALSE, message=FALSE}
ref <- MouseRNAseqData()
samples <- colData(ref)[,c("label.main", "label.fine","label.ont")]
samples <- as.data.frame(samples)
DT::datatable(unique(samples))
```

## Immune references

### Database of Immune Cell Expression (DICE)

The [Database for Immune Cell Expression(/eQTLs/Epigenomics)](https://dice-database.org) [@diceRef].

```{r, echo=FALSE, message=FALSE}
ref <- DatabaseImmuneCellExpressionData()
samples <- colData(ref)[,c("label.main", "label.fine","label.ont")]
samples <- as.data.frame(samples)
DT::datatable(unique(samples))
```

### Novershtern hematopoietic data

Novershtern Hematopoietic Cell Data - [GSE24759](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE24759) - formerly known as Differentiation Map [@dmapRef].

```{r, echo=FALSE, message=FALSE}
ref <- NovershternHematopoieticData()
samples <- colData(ref)[,c("label.main", "label.fine","label.ont")]
samples <- as.data.frame(samples)
DT::datatable(unique(samples))
```

### Monaco immune data

Monaco Immune Cell Data - [GSE107011](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE107011)  [@monaco_immuneRef].

```{r, echo=FALSE, message=FALSE}
ref <- MonacoImmuneData()
samples <- colData(ref)[,c("label.main", "label.fine","label.ont")]
samples <- as.data.frame(samples)
DT::datatable(unique(samples))
```

### Immunological Genome Project (ImmGen)

The murine [ImmGen](http://www.immgen.org/) [@ImmGenRef]

```{r, echo=FALSE, message=FALSE}
ref <- ImmGenData()
samples <- colData(ref)[,c("label.main", "label.fine","label.ont")]
samples <- as.data.frame(samples)
DT::datatable(unique(samples))
```

# Session information {-}

```{r}
sessionInfo()
```
